src/extension/handler/utils.ts

```diff
@@ -0,0 +1,139 @@
+ * we return `null` meaning we should cancel the checkout operation.
+ *
+ * @returns `vscode.Uri` with target directory to check out the project to
+ *          `null` if the user doesn't want to create the directory, here we should cancel the operation
+ */
+export async function getProjectDir () {
+    const projectDirPath: string | undefined = config.get('projectDir')
+
+    if (!projectDirPath) {
+        return Uri.parse((await dir()).path)
+    }
+
+    const projectDir = Uri.parse(projectDirPath)
+    const isExisting = await workspace.fs.stat(projectDir)
+        .then(() => true, () => false)
+    if (isExisting) {
+        return projectDir
+    }
+
+    const createDir = (await window.showInformationMessage(
+        `A project directory (${projectDir}) was set up but doesn't exist. ` +
+        'Do you want to create it?',
+        'Yes', 'No'
+    )) === 'Yes'
+
+    if (!createDir) {
+        return null
+    }
+
+    await workspace.fs.createDirectory(projectDir)
+    return projectDir
+}
+
+/**
+ * Create the name of the target directory to checkout a project to
+ * @param targetDir Uri of the base directory (received by calling `getProjectDir`)
+ * @param suggestedName name of the directory to check the project into, e.g. "org/projectName"
+ * @param index index which increases if directory name exists (e.g. "foobar_1")
+ * @returns a string with the name of the target directory
+ */
+export async function getTargetDirName (targetDir: Uri, suggestedName: string, index = 0): Promise<string> {
+    /**
+     * for now let's expect a suggested name mimicking the format "<org>/<project>"
+     */
+    if (suggestedName.split('/').length !== 2) {
+        throw new Error(`Invalid project directory suggestion: ${suggestedName}`)
+    }
+
+    /**
+     * create org directory
+     */
+    const [orgName] = suggestedName.split('/')
+    const orgDir = Uri.joinPath(targetDir, orgName)
+    const isOrgDirExisting = await workspace.fs.stat(orgDir).then(() => true, () => false)
+    if (!isOrgDirExisting) {
+        await workspace.fs.createDirectory(orgDir)
+    }
+
+    const amendedSuggestedName = !index ? suggestedName: `${suggestedName}_${index}`
+    const fullTargetDir = Uri.joinPath(targetDir, amendedSuggestedName)
+    const isExisting = await workspace.fs.stat(fullTargetDir).then(() => true, () => false)
+    if (isExisting) {
+        return getTargetDirName(targetDir, suggestedName, ++index)
+    }
+
+    return amendedSuggestedName
+}
+
+export async function waitForProjectCheckout (
+    targetDir: string,
+    repository: string,
+    cb: (success: boolean) => void
+) {
+    const targetDirUri = Uri.parse(targetDir)
+    const t = setTimeout(() => {
+        clearInterval(i)
+        window.showErrorMessage(`Timed out cloning repository ${repository}`)
+        return cb(false)
+    }, Math.max(config.get('timeout') || 0, MINIMAL_TIMEOUT))
+    const i = setInterval(async () => {
```

Comment:
If the reasons are to get something "in our/users hands" fast and improve on it later. I'm ok with this implementation. As per Max's suggestion we might want to incrementally improve if we find that it's not robust enough as part of testing/dogfooding.

Reply:
Can you explain what you mean by "more indirect"? The goal here is to wait for the checkout to be completed. The clone happens as part of a created VS Code terminal instance to which we don't have any information on neither the exit code nor the terminal stdout/err. We can surely build a more sophisticated terminal execution that would allow us to get this data but as @sourishkrout mentioned I prefer to go the scrappy route that allows us to test and iterate on this later on.

Reply:
@christian-bromann 

I had a few concerns when I wrote that. For example, the command could fail early, and then the user will get an error message 10s later. Or the repo might take more than the timeout to clone. FS polling also introduces potential race conditions such as the one I mentioned earlier. This is what I mean by "indirectness" - sorry if that was a weird description! A more sophisticated terminal execution would fix these issues, though indeed not without development expense.

Going the scrappy route is totally fine, I just wanted to bring it up since this is part of user onboarding, and it would be a shame for users to not use the product as a result of these issues. As long as we are aware of these issues though and are willing to do incremental improvements, I'm with @sourishkrout !